### PR TITLE
[Fresco] Always reserve space for a background and an overlay

### PR DESCRIPTION
**Pull request by plamenko**
This achieves two things:
* It allows us to dynamically change backgrounds and overlays
* It saves memory as we removed 5 instance integers and added only 2 instance null-references (two null elements in the array of the fade drawable)

An additional change is that multiple backgrounds are from now on deprecated. Context:
This improvement required a small interface change between GenericDraweeHierarchy and its builder. In particular, GenericDraweeHierarchy now has only one layer dedicated for a background. Multiple backgrounds are still supported via builder, but this gets emulated by wrapping them with an ArrayDrawable. In theory this should work fine, with one caveat. Rounding no longer works for multiple backgrounds. This would be relatively easy to support, but given that multiple backgrounds are rarely used feature (if ever), instead of introducing additional complexity we decided to deprecate it instead.
In case where there is only one or no background, everything should work as it was before.

Test Plan: unit tests, demo sample app

### CODE REVIEW

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java [hunk 1]
```diff
@@ -85,6 +85,14 @@
...
 public class GenericDraweeHierarchy implements SettableDraweeHierarchy {
 
+  private static final int BACKGROUND_IMAGE_INDEX = 0;
+  private static final int PLACEHOLDER_IMAGE_INDEX = 1;
+  private static final int ACTUAL_IMAGE_INDEX = 2;
+  private static final int PROGRESS_BAR_IMAGE_INDEX = 3;
+  private static final int RETRY_IMAGE_INDEX = 4;
+  private static final int FAILURE_IMAGE_INDEX = 5;
+  private static final int OVERLAY_IMAGES_INDEX = 6;
+
```
Comment (massimocarli) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: What about using @IntDef annotation here?

Comment (plamenko) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: What's the benefit?

Comment (massimocarli) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: Oh, I didn't realized they where private. Anyway it would add type safety-ish.

Comment (balazsbalazs) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: I like adding @IntDef - it's a nice touch.

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java [hunk 2]
```diff
@@ -555,26 +544,9 @@ public void setProgressBarImage(int resourceId, ScaleType scaleType) {
...
   }
 
-  /**
-   * Sets a new background image at the specified index.
-   *
-   * This method will throw if the given index is out of bounds.
-   *
-   * @param drawable background image
-   */
-  public void setBackgroundImage(int index, @Nullable Drawable drawable) {
```
Comment (balazsbalazs) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: as we discussed IRL: you just introduced this method a few days ago, right?

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java [hunk 3]
```diff
@@ -585,14 +557,10 @@ public void setBackgroundImage(@Nullable Drawable drawable) {
    * @param drawable background image
    */
   public void setOverlayImage(int index, @Nullable Drawable drawable) {
-    // Note: we do not have mOverlaysIndex to reduce the memory footprint of this class.
```
Comment (balazsbalazs) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: This two lines of comments are still relevant: we do rely on overlays being the last ones.

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchyBuilder.java [hunk 4]
```diff
@@ -581,34 +582,36 @@ public GenericDraweeHierarchyBuilder setActualImageColorFilter(@Nullable ColorFi
...
     return this;
   }
 
   /**
-   * Gets the backgrounds.
+   * Gets the background.
    */
-  public @Nullable List<Drawable> getBackgrounds() {
-    return mBackgrounds;
+  public @Nullable Drawable getBackground() {
```
Comment (balazsbalazs) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchyBuilder.java]: Ok, this is definitely a breaking change. Which is fine but let's put this in the description of the PR.
"From now on Fresco only supports one background instead of multiple ones. We've seen very little usage of multiple backgrounds and we can have better performance if we disallow multiple backgrounds"

File: samples/demo/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -110,8 +110,6 @@
       <com.facebook.drawee.view.SimpleDraweeView
         android:id="@+id/alpha_webp"
         style="@style/drawee"
-        fresco:backgroundImage="@color/transparent"
```
Comment (balazsbalazs) [inline samples/demo/src/main/res/layout/activity_main.xml]: ? Why did you remove these?

**Comment by balazsbalazs** [comment]
All in all a great change but let's do a more in-depth description.
We're basically removing support for multiple backgrounds with this commit - let's emphasize that. I would change the commit name to "Remove multiple background support" and then explain that we're doing this for perf reasons.

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java (hunk 1 above)
Comment (plamenko) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: Pushing back again :) It duplicates the definitions without a good reason. Those are index constants, not enum values.

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java (hunk 2 above)
Comment (plamenko) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchy.java]: Correct. It should be safe to remove it.

File: drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchyBuilder.java (hunk 4 above)
Comment (plamenko) [inline drawee/src/main/java/com/facebook/drawee/generic/GenericDraweeHierarchyBuilder.java]: This getter is in `GenericDraweeHierarchyBuilder`, and is most likely used exclusively by `GenericDraweeHierarchy` constructor. No one should really call this directly.

File: samples/demo/src/main/res/layout/activity_main.xml (hunk 5 above)
Comment (plamenko) [inline samples/demo/src/main/res/layout/activity_main.xml]: Because we no longer need to preallocate space for background/overlay in order to change it dynamically later on, and more importanitly we no longer need to allocate this useless transparent drawable.
Pure win :)

**Comment by plamenko** [comment]
We are not really dropping support for multiple backgrounds. There has been some internal interface changes between drawee hierarchy and its builder, but the builder setters are still there.

There is a breaking part though, which is that rounding won't work for backgrounds if there is more than one. This is easy to fix, but I didn't bother because I am reasonably confident that multiple backgrounds is rarely used if ever. I admit that was a bit ov overengineering on my side. Time to fix it :)

**Comment by facebook-github-bot-6** [comment]
@plamenko updated the pull request.
